Exploratory Data Analysis (EDA)

Project: Predicting Career Domain and Seniority from LinkedIn Profiles
Goal is to Understand the structure and characteristics of Active job positions


In [1]:
import json
from pathlib import Path
import pandas as pd


In [2]:
from pathlib import Path

PROJECT_ROOT = Path.cwd().parents[1]       # aktuelles Working Directory  #CAPSTONE-Project-/
PROJECT_ROOT



PosixPath('/Users/lennartredlich/Master/1. Semester/Practical Data Science/Capstone-Project-')

In [3]:
DATA_PATH = PROJECT_ROOT / "files"
FILE = DATA_PATH / "linkedin-cvs-annotated.json"
FILE.exists()

True

In [4]:
with open(FILE, "r", encoding="utf-8") as f:
    cvs = json.load(f)

type(cvs), len(cvs)

(list, 609)

Data Type: List
Count of CVs: 609


In [6]:
cvs[0]

[{'organization': 'Depot4Design GmbH',
  'linkedin': 'https://www.linkedin.com/company/depot4design-gmbh',
  'position': 'Prokurist',
  'startDate': '2019-08',
  'endDate': None,
  'status': 'ACTIVE',
  'department': 'Other',
  'seniority': 'Management'},
 {'organization': 'Depot4Design GmbH',
  'linkedin': 'https://www.linkedin.com/company/depot4design-gmbh',
  'position': 'CFO',
  'startDate': '2019-07',
  'endDate': None,
  'status': 'ACTIVE',
  'department': 'Other',
  'seniority': 'Management'},
 {'organization': 'Depot4Design GmbH',
  'linkedin': 'https://www.linkedin.com/company/depot4design-gmbh',
  'position': 'Betriebswirtin',
  'startDate': '2019-07',
  'endDate': None,
  'status': 'ACTIVE',
  'department': 'Other',
  'seniority': 'Professional'},
 {'organization': 'Depot4Design GmbH',
  'linkedin': 'https://www.linkedin.com/company/depot4design-gmbh',
  'position': 'Prokuristin',
  'startDate': '2019-07',
  'endDate': None,
  'status': 'ACTIVE',
  'department': 'Other',
  '

In [9]:
from collections import Counter

active_job_counts = []

for cv in cvs:
    count = sum(1 for job in cv if job.get("status") == "ACTIVE")
    active_job_counts.append(count)

Counter(active_job_counts)


Counter({1: 380, 0: 131, 2: 78, 3: 10, 5: 4, 4: 3, 8: 1, 7: 1, 10: 1})

Most individuals have 1 job, second most individuals have zero active jobs and so on...




In [8]:
import pandas as pd

jobs = []
for cv in cvs:
    for job in cv:
        jobs.append(job)

df = pd.DataFrame(jobs)
df.head()





,organization,linkedin,position,startDate,endDate,status,department,seniority
0,Depot4Design GmbH,https://www.linkedin.com/company/depot4design-...,Prokurist,2019-08,None,ACTIVE,Other,Management
1,Depot4Design GmbH,https://www.linkedin.com/company/depot4design-...,CFO,2019-07,None,ACTIVE,Other,Management
2,Depot4Design GmbH,https://www.linkedin.com/company/depot4design-...,Betriebswirtin,2019-07,None,ACTIVE,Other,Professional
3,Depot4Design GmbH,https://www.linkedin.com/company/depot4design-...,Prokuristin,2019-07,None,ACTIVE,Other,Management
4,Depot4Design GmbH,https://www.linkedin.com/company/depot4design-...,CFO,2019-07,None,ACTIVE,Other,Management


In [10]:
df_active = df[df["status"] == "ACTIVE"]
df_active.shape

(623, 8)

There are 623 jobs with the Status Active currently.

In [11]:
df[["department", "seniority"]].isna().mean()


department    0.0
seniority     0.0
dtype: float64

All job entries in the data set are fully labeled with both department and seniority information. No missing Values. 

In [12]:
df["department"].value_counts(normalize=True)


department
Other                     0.474602
Information Technology    0.118271
Sales                     0.083017
Consulting                0.073920
Project Management        0.066338
Marketing                 0.050417
Administrative            0.031842
Business Development      0.029568
Purchasing                0.027293
Human Resources           0.026535
Customer Support          0.018196
Name: proportion, dtype: float64

Okay department is fully labeled, but as you can see The Label "Other" makes 47% of all Entries. 


In [13]:
df[df["department"] == "Other"]["position"].value_counts().head(15)


position
Geschäftsführer                      41
Managing Director                    22
CEO                                  20
Owner                                12
Director                             11
Vice President                       11
General Manager                      10
Geschäftsführender Gesellschafter     9
Unternehmensinhaber                   9
Rechtsanwalt                          8
Professor                             8
CFO                                   7
Research Assistant                    7
Intern                                7
Chief Executive Officer               6
Name: count, dtype: int64

In [14]:
df["seniority"].value_counts(normalize=True)

seniority
Professional    0.462092
Lead            0.174375
Management      0.158453
Junior          0.087187
Senior          0.064443
Director        0.053450
Name: proportion, dtype: float64

So as you can see there the seniority is complete. 

In [15]:
with open(DATA_PATH / "linkedin-cvs-not-annotated.json", "r", encoding="utf-8") as f:
    cvs_unlabeled = json.load(f)

len(cvs_unlabeled)

390

In [16]:
cvs_unlabeled[0]

[{'organization': 'Keeping The Books, Bookkeeping',
  'linkedin': '',
  'position': 'Bookkeeper',
  'startDate': '2023-03',
  'endDate': None,
  'status': 'ACTIVE'},
 {'organization': 'Playful Paws',
  'linkedin': '',
  'position': 'Co-Owner',
  'startDate': '2018-11',
  'endDate': None,
  'status': 'ACTIVE'},
 {'organization': 'S&R services',
  'linkedin': '',
  'position': 'Logistics Officer',
  'startDate': '2019-09',
  'endDate': '2024-04',
  'status': 'INACTIVE'},
 {'organization': 'ABC Supply Co. Inc.',
  'linkedin': 'https://www.linkedin.com/company/abc-supply',
  'position': 'Truck driver/ laborer',
  'startDate': '2019-03',
  'endDate': '2019-09',
  'status': 'INACTIVE'},
 {'organization': 'MB Railways',
  'linkedin': '',
  'position': 'Fuel Driver',
  'startDate': '2018-03',
  'endDate': '2019-03',
  'status': 'INACTIVE'},
 {'organization': 'Sysco',
  'linkedin': 'https://www.linkedin.com/company/sysco',
  'position': 'Food Delivery Driver',
  'startDate': '2017-10',
  'endDa

The unannotated dataset follows the same structural schema as the annotated CVs

In [20]:
pd.read_csv(DATA_PATH / "department-v2.csv").head()

,text,label
0,Adjoint directeur communication,Marketing
1,Advisor Strategy and Projects,Project Management
2,Beratung & Projekte,Project Management
3,Beratung & Projektmanagement,Project Management
4,Beratung und Projektmanagement kommunale Partner,Project Management


In [28]:
df_department_v2_labels = pd.read_csv(DATA_PATH / "department-v2.csv")
df_department_v2_labels["label"].nunique()



11

In [29]:
df_department_v2_labels["label"].value_counts()

label
Marketing                 4295
Sales                     3328
Information Technology    1305
Business Development       620
Project Management         201
Consulting                 167
Administrative              83
Other                       42
Purchasing                  40
Customer Support            33
Human Resources             31
Name: count, dtype: int64

In [30]:
df_seniority_v2 = pd.read_csv(DATA_PATH / "seniority-v2.csv")
df_seniority_v2["label"].nunique()

5

In [32]:
df_seniority_v2["label"].unique()

array(['Junior', 'Senior', 'Lead', 'Management', 'Director'], dtype=object)

In [31]:
df_seniority_v2["label"].value_counts() 

label
Senior        3733
Lead          3546
Director       984
Management     756
Junior         409
Name: count, dtype: int64

The seniority taxonomy defined in 'seniority-v2.csv' consists of five levels: Junior, Senior, Lead, Management, and Director. 

In comparison with the annotated LinkedIn CVs, we observe that the label "Professional" appears in the CV data but is not part of the predefined seniority label. 
